docs/KubeletConfigDesign.md

```diff
@@ -0,0 +1,119 @@
+
+4. setting max pods per node ([https://docs.openshift.com/container-platform/3.11/admin_guide/manage_nodes.html#admin-guide-max-pods-per-node](https://docs.openshift.com/container-platform/3.11/admin_guide/manage_nodes.html#admin-guide-max-pods-per-node))
+
+5. configuring garbage collection ([https://docs.openshift.com/container-platform/3.11/admin_guide/garbage_collection.html](https://docs.openshift.com/container-platform/3.11/admin_guide/garbage_collection.html))
+
+6. configuring node resources ([https://docs.openshift.com/container-platform/3.11/admin_guide/manage_nodes.html#configuring-node-resources](https://docs.openshift.com/container-platform/3.11/admin_guide/manage_nodes.html#configuring-node-resources))
+
+## Non-Goals
+
+# Proposal
+
+Extend the Machine Config Operator to include a KubetletConfig CRD and KubeletConfigController. By using a KubeletConfig CRD there is an implicit whitelist of allowed user controlled options. Upon deleting the KubeletConfig instance the default config is restored.
+
+## CRD
+
+```
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: kubeletconfigs.machineconfiguration.openshift.io
+spec:
+  group: machineconfiguration.openshift.io
+  versions:
+    - name: v1
+      served: true
+      storage: true
+  scope: Cluster
+  names:
+    plural: kubeletconfigs
+    singular: kubeletconfig
+    kind: KubeletConfig
+```
+
+## SPEC
+
+```
+MachineConfigPoolSelector *metav1.LabelSelector
+Runtime:
+  Name  string
+  Endpoint string
+KubeletConfig
+  [KubeletConfigurationSpec](https://github.com/kubernetes/kubernetes/blob/release-1.11/pkg/kubelet/apis/kubeletconfig/v1beta1/types.go#L45)
+```
+
+Example:
+
+```
+apiVersion: machineconfiguration.openshift.io/v1
+kind: KubeletConfig
+metadata:
+  name: set-max-pods
+spec:
+  machineConfigPoolSelector:
+    matchLabels:
+      custom-kubelet: small-pods
+  kubeletConfig:
+    maxPods: 100
+```
+
+## Implementation Details
+
+The KubeletConfigController would perform the following steps:
+
+1. Validate the user defined KubeletConfig
+
+2. Render the current MachineConfig (storage.files.contents[kubelet.conf]) into the KubeletConfiguration structure
+
+3. Load the KubeletConfig from the passed in Spec.KubeletConfig
+
+4. Use mergo to merge the two structures
+
+5. Serialize the KubeletConfig to yaml
+
+6. Create or Update an ignition /etc/kubernetes/kubelet.conf file within a 99-[role]-kubelet-managed MachineConfig
+
+After deletion of the KubeletConfig instance the config will be reverted to the original kubelet config.
+
+## Runtime Selection
+
+### Requirements
+
+First time installs will **always** install with CRI-O. After the initial install, the user will be able to switch from CRI-O to Docker, or from Docker to CRI-O if commanded to.
```

Comment:
Really?

Reply:
@sjenning This is something Seth is following. I'm not sure how 'real' it is.

Reply:
last I heard from @derekwaynecarr , this was the plan.